File: 702_Search_in_a_Sorted_Array_of_Unknown_Size.py
# Time Complexity : O(logn)
# 
# Space Complexity : O(1)
# 
# Did this code successfully run on Leetcode : Yes
# 
# Any problem you faced while coding this : No
#
# Problem Approach
# 1. Since the input list is an infinite sorted steam, we need to define the end index to apply binary search. Any index which holds a value greater or equal to the target can be taken as end index
# 2. The end index is initialized with value 1. The value is doubled at each iteration until a desired value has been reached





# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        low, high = 0,1
        
        # Look for the end index
        while reader.get(high) < target:
            low = high + 1
            high *= 2
            
        while low<= high:
            mid = (low+high) // 2
            if reader.get(mid) == target:
                return mid
            elif reader.get(mid) < target:
                low = mid+1
            else:
                high = mid-1
        return -1

File: 702. Search in a Sorted Array of Unknown Size.cpp
//Time Complexity = O(log n)
//Space Complexity = O(1)

/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution {
public:
    int BinarySearch(ArrayReader reader, int l, int h, int target)
    {
        int low = l, high = h, mid;
        while(low <= high)
        {
            mid = low + (high - low)/2;
            if(reader.get(mid) == target)
                return mid;
            else if(reader.get(mid) < target)
                low = mid + 1;
            else
                high = mid -1;
        }
        return -1;
    }
    int search(const ArrayReader& reader, int target) {
        int low = 0, high = 1;
        while(reader.get(high) < target)
        {
            low = high;
            high = 2 * low;
        }
        return BinarySearch(reader, low, high, target);
    }
};
